File: 351001/Boukhvalova/module-publisher/src/main/java/by/boukhvalova/distcomp/controllers/v2/StickerV2Controller.java
package by.boukhvalova.distcomp.controllers.v2;

import by.boukhvalova.distcomp.dto.StickerRequestTo;
import by.boukhvalova.distcomp.dto.StickerResponseTo;
import by.boukhvalova.distcomp.services.StickerService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;

@RestController
@RequestMapping({"/api/v2.0/stickers", "/api/v2.0/markers", "/api/v2.0/labels"})
@RequiredArgsConstructor
public class StickerV2Controller {
    private final StickerService stickerService;

    @GetMapping
    public Collection<StickerResponseTo> getAll() {
        return stickerService.getAll();
    }

    @GetMapping("/{id}")
    public StickerResponseTo getById(@PathVariable Long id) {
        return stickerService.getById(id);
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    @PreAuthorize("hasRole('ADMIN')")
    public StickerResponseTo create(@RequestBody @Valid StickerRequestTo request) {
        return stickerService.create(request);
    }

    @PutMapping
    @ResponseStatus(HttpStatus.OK)
    @PreAuthorize("hasRole('ADMIN')")
    public StickerResponseTo update(@RequestBody @Valid StickerRequestTo request) {
        return stickerService.update(request);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @PreAuthorize("hasRole('ADMIN')")
    public void delete(@PathVariable Long id) {
        stickerService.delete(id);
    }
}

File: 351002/Abrashin/publisher/src/main/kotlin/com/example/distcomp/controller/StickerV2Controller.kt
package com.example.distcomp.controller

import com.example.distcomp.dto.request.StickerRequestTo
import com.example.distcomp.dto.response.StickerResponseTo
import com.example.distcomp.security.AuthorizationService
import com.example.distcomp.service.StickerService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v2.0/stickers")
class StickerV2Controller(
    private val service: StickerService,
    private val authorizationService: AuthorizationService
) {
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(@Valid @RequestBody request: StickerRequestTo): StickerResponseTo {
        authorizationService.ensureAdmin()
        return service.create(request)
    }

    @GetMapping("/{id}")
    fun getById(@PathVariable id: Long): StickerResponseTo = service.getById(id)

    @GetMapping
    fun getAll(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
        @RequestParam(defaultValue = "id,desc") sort: Array<String>
    ): List<StickerResponseTo> = service.getAll(page, size, sort)

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @Valid @RequestBody request: StickerRequestTo): StickerResponseTo {
        authorizationService.ensureAdmin()
        return service.patch(id, request)
    }

    @PatchMapping("/{id}")
    fun patch(@PathVariable id: Long, @Valid @RequestBody request: StickerRequestTo): StickerResponseTo {
        authorizationService.ensureAdmin()
        return service.patch(id, request)
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun delete(@PathVariable id: Long) {
        authorizationService.ensureAdmin()
        service.delete(id)
    }
}
